cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,232 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.cats.redis.cluster;
+
+import com.netflix.spinnaker.cats.agent.*;
+import com.netflix.spinnaker.cats.module.CatsModuleAware;
+import com.netflix.spinnaker.cats.redis.JedisSource;
+import com.netflix.spinnaker.cats.thread.NamedThreadFactory;
+import redis.clients.jedis.Jedis;
+
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/*
+ * The idea behind this scheduler is simple. Every agent it owns is always in one of two sorted sets,
+ * WORKING, or WAITING. If it is in the WORKING set, it's being run, if it's in the WAITING set, it is waiting to run.
+ * These sets are sorted by time, and the rank is time at which they are ready to run (in the WAITING set), or have
+ * timed out (in the WORKING set). Therefore, if we do a `zrangebyscore SETNAME -inf CURRENT_TIME`, we get a list
+ * of all agents that have expired. Those are then atomically pulled from the set, and run by the scheduler.
+ *
+ * The amortized cost of this scheduler is much lower than the original ClusteredAgentScheduler, since during each
+ * cache interval every key will only be removed from Redis once. If the interval is 60s, and the agent polls every 1s,
+ * we already have a (30s / 1) * (# of clouddrivers) factor of improvement.
+ */
+class ClusteredSortAgentScheduler extends CatsModuleAware implements AgentScheduler, Runnable {
+  private final JedisSource jedisSource;
+  private final NodeStatusProvider nodeStatusProvider;
+  private final AgentIntervalProvider intervalProvider;
+  private final ExecutorService agentWorkPool;
+  private final Integer parallelism;
+  private Integer runCount = 0;
+  private final Integer REDIS_REFRESH_PERIOD = 30; // How many iterations to wait between repopulating redis with potentially missing keys
+
+  private static Integer NOW = 0;
+
+  private Map<String, AgentWorker> agents;
+  private Semaphore runningAgents;
+
+  // This code assumes that every agent being run is in exactly either the WAITING or WORKING set.
+  private static final String WAITING_SET = "WAITZ";
+  private static final String WORKING_SET = "WORKZ";
+  private static final String ADD_AGENT_SCRIPT = "addAgentScript";
+  private static final String SWAP_SET_SCRIPT = "swapSetScript";
+
+  private HashMap<String, String> scriptShas;
+
+  public ClusteredSortAgentScheduler(JedisSource jedisSource, NodeStatusProvider nodeStatusProvider, AgentIntervalProvider intervalProvider, Integer parallelism) {
+    this.jedisSource = jedisSource;
+    this.nodeStatusProvider = nodeStatusProvider;
+    this.agents = new HashMap<>();
```

Comment:
This should be ConcurrentHashMap (accessed during run and during schedule calls)